### PR TITLE
Update build to Gradle 7.0

### PR DESCRIPTION
**Pull request by breskeby**
Gradle 7.0 GA has been released. We want to stay on the latest released Gradle version for our elasticsearch builds. 

This updates our build to use Gradle 7.0 and applies some required changes to make our build compatible with the latest Gradle version.

Most of the heavy lifting to support 7.0 have been merged into master already in earlier separated PRs.

As this is a bigger change I also triggered a few other CI builds to get more reassurance than usual:

- https://elasticsearch-ci.elastic.co/view/Elasticsearch%20master/job/elastic+elasticsearch+master+periodic/65/
- https://elasticsearch-ci.elastic.co/view/Elasticsearch%20master/job/elastic+elasticsearch+master+third-party-tests/111/
- https://elasticsearch-ci.elastic.co/view/Elasticsearch%20master/job/elastic+elasticsearch+master+packaging-gcp/193/

### CODE REVIEW

**Comment by breskeby** [comment]
currently blocked by https://github.com/gradle/gradle/issues/15566

**Comment by breskeby** [comment]
currently blocked by https://github.com/nebula-plugins/gradle-ospackage-plugin/issues/393

File: build.gradle
Comment (breskeby) [inline build.gradle:286]: deprecated `compile` has been removed. 
This should have been replaced before but now ultimately caused the build to fail

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/doc/SnippetsTask.groovy
Comment (breskeby) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/doc/SnippetsTask.groovy:33]: $SKIP seems like a reserved term in Groovy 3.0 and caused trouble here. Therefore the renaming

File: buildSrc/src/main/java/org/elasticsearch/gradle/RepositoriesSetupPlugin.java
Comment (breskeby) [inline buildSrc/src/main/java/org/elasticsearch/gradle/RepositoriesSetupPlugin.java:41]: Resolving this TODO by removing the extra logic for http checks

File: buildSrc/src/main/java/org/elasticsearch/gradle/test/TestWithSslPlugin.java
Comment (breskeby) [inline buildSrc/src/main/java/org/elasticsearch/gradle/test/TestWithSslPlugin.java:46]: Make implicit dependencies explicit

File: buildSrc/src/test/java/org/elasticsearch/gradle/internal/precommit/DependencyLicensesTaskTests.java
Comment (breskeby) [inline buildSrc/src/test/java/org/elasticsearch/gradle/internal/precommit/DependencyLicensesTaskTests.java:59]: Groovy is shipped with different artefacts now

File: gradle/formatting.gradle
Comment (breskeby) [inline gradle/formatting.gradle:166]: narrow down the tasks inputs here to avoid implicit dependencies on others tasks output we do not want here (and is deprecated with gradle 7.0)

File: plugins/repository-hdfs/build.gradle
Comment (breskeby) [inline plugins/repository-hdfs/build.gradle:72]: make task dependency explicit

File: x-pack/plugin/ml/build.gradle
Comment (breskeby) [inline x-pack/plugin/ml/build.gradle:83]: use a dedicated output folder for extracted native licenses. This avoids not having the whole `buildDir` considered a task output here

**Comment by breskeby** [comment]
@elasticmachine run elasticsearch-ci/2

**Comment by breskeby** [comment]
I'll look into elasticsearch-hadoop build first before merging this to ensure elasticsearch-hadoop is ready to be updated to 7.0 too

**Comment by breskeby** [comment]
Elastic-hadoop PR has been created here: https://github.com/elastic/elasticsearch-hadoop/pull/1641

File: distribution/packages/build.gradle
Comment (mark-vieira) [inline distribution/packages/build.gradle:51]: Why do we need to fork this?

**Review by mark-vieira — APPROVED** [review]
Woohoo! LGTM.

File: distribution/packages/build.gradle
Comment (breskeby) [reply distribution/packages/build.gradle:51]: I'm waiting for a release of the ospackage plugin that includes my PR that fixed a stackoverflow within the plugin: https://github.com/nebula-plugins/gradle-ospackage-plugin/pull/397

**Pull request merged**
